Source/WebCore/loader/ContentFilter.cpp

```diff
@@ -123,6 +123,76 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+{
+    ASSERT(RunLoop::isMain());
+    ASSERT(m_callback);
+
+    if (m_isBlocked) {
+        if (CheckedPtr contentFilter = m_contentFilter)
+            contentFilter->didDecide(State::Blocked);
+        m_callback(ResourceRequest());
+        return;
+    }
+
+    if (CheckedPtr contentFilter = m_contentFilter; m_numberOfFiltersAllowed == contentFilter->m_contentFilters.size())
+        contentFilter->didDecide(State::Allowed);
+
+    m_callback(WTFMove(m_request));
+}
+
+void ContentFilter::ContentFilterCallbackAggregator::didReceivePlatformContentFilterDecision(PlatformContentFilter& platformContentFilter, ResourceRequest&& request, PlatformContentFilter::State state)
+{
+    ASSERT(RunLoop::isMain());
+
+    if (state == PlatformContentFilter::State::Allowed)
+        ++m_numberOfFiltersAllowed;
+
+    bool isBlocked = state == PlatformContentFilter::State::Blocked;
+    if (isBlocked && !m_isBlocked) {
+        m_isBlocked = true;
+        CheckedPtr contentFilter = m_contentFilter;
+        contentFilter->m_blockingContentFilter = platformContentFilter;
+        return;
+    }
+
+    m_request = WTFMove(request);
+}
+
+ContentFilter::ContentFilterCallbackAggregator::ContentFilterCallbackAggregator(ContentFilter& contentFilter, CompletionHandler<void(ResourceRequest&&)>&& callback)
+    : m_contentFilter(contentFilter)
+    , m_callback(WTFMove(callback))
+{
+    ASSERT(RunLoop::isMain());
+}
+
+void ContentFilter::continueAfterWillSendRequest(ResourceRequest&& request, const ResourceResponse& redirectResponse, CompletionHandler<void(ResourceRequest&&)>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+
+    // Return early if no platform content filters are enabled.
+    if (!anyContentFilterIsEnabled()) {
+        continueAfterWillSendRequest(request, redirectResponse);
+        completionHandler(WTFMove(request));
+        return;
+    }
+
+    LOG(ContentFiltering, "ContentFilter received request for <%{sensitive}s> with redirect response from <%{sensitive}s>.\n", request.url().string().utf8().data(), redirectResponse.url().string().utf8().data());
+    ASSERT(m_state == State::Stopped || m_state == State::Filtering);
+
+    Ref contentFilterCallbackAggregator = ContentFilterCallbackAggregator::create(*this, WTFMove(completionHandler));
+
+    for (Ref platformContentFilter : m_contentFilters) {
+        CompletionHandler<void(ResourceRequest, PlatformContentFilter::State)> completion = [contentFilterCallbackAggregator, platformContentFilter] (ResourceRequest&& request, PlatformContentFilter::State state) mutable {
+            ASSERT(RunLoop::isMain());
+            contentFilterCallbackAggregator->didReceivePlatformContentFilterDecision(platformContentFilter, WTFMove(request), state);
+        };
+
+        ASSERT(platformContentFilter->needsMoreData());
+        platformContentFilter->willSendRequest(ResourceRequest { request }, redirectResponse, WTFMove(completion));
```

Comment:
In the existing implementation, I think we use updated request to query the next `PlatformContentFilter`; but in the new implementation we ask all `PlatformContentFilter`s with the original request. I am not sure if this is expected.